### ISSUE TITLE
Bug with encryption block and `remote_state_data_sources`

### ISSUE THREAD

**Issue by Evi1Pumpkin** [opentofu/opentofu#1978]
### OpenTofu Version

```shell
1.8.2
```

### OpenTofu Configuration Files

I have two tofu configurations in the directories `./another`, `./root`:
:one: :
```hcl
// ./another/main.tf

variable "passphrase_another" {
  default = "changemepleaseihopeimlongenough"
}

terraform {
  encryption {
    key_provider "pbkdf2" "mykey" {
      passphrase = var.passphrase_another
    }
    method "aes_gcm" "new_method" {
      keys = key_provider.pbkdf2.mykey
    }

    state {
      method = method.aes_gcm.new_method
      enforced = true
    }
  }
}

resource "null_resource" "foo" {
}

output "test" {
   value = null_resource.foo.id
}
```

:two: :
```hcl
// ./root/main.tf:

variable "passphrase_root" {
  default = "changemepleaseihopeimlongenoughroot"
}

variable "passphrase_another" {
  default = "changemepleaseihopeimlongenough"
}

terraform {
  encryption {
    key_provider "pbkdf2" "mykey" {
      passphrase = var.passphrase_root
    }
    method "aes_gcm" "new_method" {
      keys = key_provider.pbkdf2.mykey
    }

    // key and method for the remote_state_data_sources
    key_provider "pbkdf2" "another_key" {
      passphrase = var.passphrase_another
    }
    method "aes_gcm" "another_method" {
      keys = key_provider.pbkdf2.another_key
    }

    state {
      method = method.aes_gcm.new_method
      enforced = true
    }

    remote_state_data_sources {
      default {
        method = method.aes_gcm.another_method
      }
    }
  }
}

resource "null_resource" "a" {
}

// data source to read another state
data "terraform_remote_state" "another_state" {
  backend = "local"

  config = {
   path = "../another/terraform.tfstate"
  }
}

resource "local_file" "foo" {
  content  = "hello"
  filename = "${data.terraform_remote_state.another_state.outputs.test}.txt"
}
```

### Debug Output

```
Acquiring state lock. This may take a few moments...
null_resource.a: Refreshing state... [id=520290251521991006]
data.terraform_remote_state.another_state: Reading...

Planning failed. OpenTofu encountered an error while generating this plan.

╷
│ Error: decryption failed for all provided methods: attempted decryption failed for remote.default: no decryption key available
│
│
╵
╷
│ Error: data.terraform_remote_state.another_state: Unable to read remote state
│
```

### Expected Behavior

Should successfully apply in the `root` directory.

### Actual Behavior

I get an error.

### Steps to Reproduce

1. Go to `./another`:
`tofu init`
`tofu apply`
Look at the state and make sure it's encrypted.

2. Go to `./root`:
`tofu init`
`tofu apply`
Get an error.

**Comment by cam72cam** [comment]
Ok, this is one of the hard corners on how we setup encryption.  This is a duplicate of https://github.com/opentofu/opentofu/issues/1605

**Comment by yottta** [comment]
Adding some context on why this issue has been solved by the solution provided for #1605:

Because the way meta information of the used key is saved, in the configuration where a remote state is wanted to be read, `encrypted_metadata_alias` can be used to configure the same meta key that has been used to write the encrypted state.

Therefore, in order to let OpenTofu know what key provider meta key should be used, there are two scenarios:
* `another` configuration `key_provider` has **no** `encrypted_metadata_alias` defined
* `another` configuration `key_provider` has `encrypted_metadata_alias` defined

For the first situation, if the `another` configuration `key_provider` contains no `encrypted_metadata_alias`, the `root` configuration will have to use the fully qualified name of the `key_provider` used in `another`:
```hcl
// another/main.tf
terraform {
  encryption {
    key_provider "pbkdf2" "mykey" {
      passphrase = var.passphrase_another
    }
    // ...
  }
  // ...
}
```
```hcl
// root/main.tf
terraform {
  encryption {
    // ...
    // key and method for the remote_state_data_sources
    key_provider "pbkdf2" "another_key" {
      passphrase = var.passphrase_another
      encrypted_metadata_alias = "key_provider.pbkdf2.mykey" // <-- here we specified the fully qualified name of the key_provider in `another/main.tf`.
    }
    // ...
  }
}
```

For the second situation, when the `another` configuration `key_provider` is configured with `encrypted_metadata_alias`, the `root` configuration will have to use exactly the same value for the `key_provider` configured to read `another`'s state:
```hcl
// another/main.tf
terraform {
  encryption {
    key_provider "pbkdf2" "mykey" {
      passphrase = var.passphrase_another
      encrypted_metadata_alias = "my_custom_meta_alias" // <-- configure what metadata key should be used for the key information
    }
    // ...
  }
  // ...
}
```
```hcl
// root/main.tf
terraform {
  encryption {
    // ...
    // key and method for the remote_state_data_sources
    key_provider "pbkdf2" "another_key" {
      passphrase = var.passphrase_another
      encrypted_metadata_alias = "my_custom_meta_alias" // <-- configure the same value that was used in `another/main.tf`.
    }
    // ...
  }
}
```

---

This is documented in the page for [state encryption](https://opentofu.org/docs/language/state/encryption/#remote-state-data-sources).

**Comment by brettcurtis** [comment]
Thank you @yottta ! This is very helpful and works as expected.
